# ImpPres LLM Baseline

You have to implement in this notebook a baseline for ImpPres classification using an LLM.
This baseline must be implemented using DSPy.



In [35]:
# Configure the DSPy environment with the language model - for grok the parameters must be:
# env variable should be in os.environ['XAI_API_KEY']
# "xai/grok-3-mini"
from os import environ
from os.path import exists
from datetime import datetime
import dspy
import pandas as pd

import logging
logging.getLogger("dspy.adapters.json_adapter").setLevel(logging.ERROR)

lm = dspy.LM('xai/grok-3-mini', api_key=environ['XAI_API_KEY'])

# for ollama
# lm = dspy.LM('ollama_chat/devstral', api_base='http://localhost:11434', api_key='')
# lm = dspy.LM(
#     "ollama/llama3.1:8b",
#     api_base="http://localhost:11434",
#     format="json"        # litellm translates this to Ollama's stream=false
# )
dspy.configure(lm=lm)

In [36]:
from typing import Literal

## Implement the DSPy program to classify pairs (premise, hypothesis) as entailment, contradiction, or neutral.
class NLIImPresClassifier(dspy.Signature):
    premise     :str = dspy.InputField(desc="A short passage or statement. All facts should be inferred from this text alone.")
    hypothesis  :str = dspy.InputField(desc="A second statement to evaluate. Check if this follows from, contradicts, or is unrelated to the premise.")
    label       : Literal["entailment", "neutral", "contradiction"] = dspy.OutputField(
        desc=(
            "Return one of: 'entailment', 'neutral', or 'contradiction'.\n"
            "- 'entailment': The hypothesis must be true if the premise is true.\n"
            "- 'contradiction': The hypothesis must be false if the premise is true.\n"
            "- 'neutral': The hypothesis could be either true or false based on the premise."
        )
    )

predictor = dspy.Predict(NLIImPresClassifier)
label_names = ["entailment", "neutral", "contradiction"]
def zero_shot_nli_classifier(x):
    return {
        'premise' : x['premise'],
        'hypothesis': x['hypothesis'],
        'pred_label' : predictor(premise=x['premise'], hypothesis=x['hypothesis']).label,
        'gold_label' : label_names[x['gold_label']]
    }

## Load ImpPres dataset

In [37]:
from datasets import load_dataset, Dataset
import pandas as pd
from os.path import exists

# Define sections
sections = [
    'presupposition_all_n_presupposition',
    'presupposition_both_presupposition',
    'presupposition_change_of_state',
    'presupposition_cleft_existence',
    'presupposition_cleft_uniqueness',
    'presupposition_only_presupposition',
    'presupposition_possessed_definites_existence',
    'presupposition_possessed_definites_uniqueness',
    'presupposition_question_presupposition'
]

dataset = {}

if not exists('combined_imppres_presuppositions.parquet'):
    # Load each section
    for section in sections:
        print(f"Loading dataset for section: {section}")
        dataset[section] = load_dataset("facebook/imppres", section)

    # Convert to dataframes and combine
    dataframes_list = []
    for section, data in dataset.items():
        df = data.to_pandas()
        df['section'] = section
        dataframes_list.append(df)

    combined_df = pd.concat(dataframes_list, ignore_index=True)

else:
    combined_df = pd.read_parquet('combined_imppres_presuppositions.parquet')
    print(f"Loaded combined_imppres_presuppositions.parquet")

# Convert back to datasets
dataset = {}
for section, group in combined_df.groupby("section"):
    dataset[section] = Dataset.from_pandas(group)

Loaded combined_imppres_presuppositions.parquet


In [38]:
dataset

{'presupposition_all_n_presupposition': Dataset({
     features: ['premise', 'hypothesis', 'trigger', 'trigger1', 'trigger2', 'presupposition', 'gold_label', 'UID', 'pairID', 'paradigmID', 'section', '__index_level_0__'],
     num_rows: 1900
 }),
 'presupposition_both_presupposition': Dataset({
     features: ['premise', 'hypothesis', 'trigger', 'trigger1', 'trigger2', 'presupposition', 'gold_label', 'UID', 'pairID', 'paradigmID', 'section', '__index_level_0__'],
     num_rows: 1900
 }),
 'presupposition_change_of_state': Dataset({
     features: ['premise', 'hypothesis', 'trigger', 'trigger1', 'trigger2', 'presupposition', 'gold_label', 'UID', 'pairID', 'paradigmID', 'section', '__index_level_0__'],
     num_rows: 1900
 }),
 'presupposition_cleft_existence': Dataset({
     features: ['premise', 'hypothesis', 'trigger', 'trigger1', 'trigger2', 'presupposition', 'gold_label', 'UID', 'pairID', 'paradigmID', 'section', '__index_level_0__'],
     num_rows: 1900
 }),
 'presupposition_cleft_

In [39]:
display(combined_df)

premise  \
0      All ten guys that proved to boast were divorcing.   
1      All ten guys that proved to boast were divorcing.   
2      All ten guys that proved to boast were divorcing.   
3      All ten guys that proved to boast weren't divo...   
4      All ten guys that proved to boast weren't divo...   
...                                                  ...   
17095  If the actors do conceal where that mall shock...   
17096  The actors didn't conceal where that mall shoc...   
17097  Did the actors conceal where that mall shocks ...   
17098  The actors might have concealed where that mal...   
17099  If the actors do conceal where that mall shock...   

                                              hypothesis         trigger  \
0       There are exactly ten guys that proved to boast.      unembedded   
1      There are exactly eleven guys that proved to b...      unembedded   
2      There are exactly ten senators that proved to ...      unembedded   
3       There are exactly ten guys that proved to boast.         negated   
4      There are exactly eleven guys that proved to b...         negated   
...                                                  ...             ...   
17095                               Travel shocks Janet.     conditional   
17096  The actors do conceal where that mall shocks J...  Not_In_Example   
17097  The actors do conceal where that mall shocks J...  Not_In_Example   
17098  The actors do conceal where that mall shocks J...  Not_In_Example   
17099  The actors do conceal where that mall shocks J...  Not_In_Example   

             trigger1        trigger2  presupposition  gold_label  \
0      Not_In_Example  Not_In_Example        positive           0   
1      Not_In_Example  Not_In_Example         negated           2   
2      Not_In_Example  Not_In_Example         neutral           1   
3      Not_In_Example  Not_In_Example        positive           0   
4      Not_In_Example  Not_In_Example         negated           2   
...               ...             ...             ...         ...   
17095  Not_In_Example  Not_In_Example         neutral           1   
17096         negated      unembedded  Not_In_Example           2   
17097   interrogative      unembedded  Not_In_Example           1   
17098           modal      unembedded  Not_In_Example           1   
17099     conditional      unembedded  Not_In_Example           1   

                           UID pairID  paradigmID  \
0         all_n_presupposition     0e           0   
1         all_n_presupposition     1c           0   
2         all_n_presupposition     2n           0   
3         all_n_presupposition     3e           0   
4         all_n_presupposition     4c           0   
...                        ...    ...         ...   
17095  question_presupposition  1895n          99   
17096  question_presupposition  1896c          99   
17097  question_presupposition  1897n          99   
17098  question_presupposition  1898n          99   
17099  question_presupposition  1899n          99   

                                      section  
0         presupposition_all_n_presupposition  
1         presupposition_all_n_presupposition  
2         presupposition_all_n_presupposition  
3         presupposition_all_n_presupposition  
4         presupposition_all_n_presupposition  
...                                       ...  
17095  presupposition_question_presupposition  
17096  presupposition_question_presupposition  
17097  presupposition_question_presupposition  
17098  presupposition_question_presupposition  
17099  presupposition_question_presupposition  

[17100 rows x 11 columns]

## Evaluate Metrics

Let's use the huggingface `evaluate` package to compute the performance of the baseline.


In [77]:
import evaluate
clf_metrics = evaluate.combine(["accuracy", "f1", "precision", "recall"])

## Your Turn

Compute the classification metrics on the baseline LLM model on each test section of the ANLI dataset for samples that have a non-empty 'reason' field.

You also must show a comparison between the DeBERTa baseline model and this LLM baseline model. The comparison metric should compute the agreement between the two models:
* On how many samples they are both correct [Correct]
* On how many samples Model1 is correct and Model2 is incorrect [Correct1]
* On how many samples Model1 is incorrect and Model2 is correct [Correct2]
* On how many samples both are incorrect [Incorrect]

We will first run the dspy classifier through the dataset:

In [41]:
def accuracy_metric(example, pred, *args):
     return pred.label == example.label

In [42]:
import pandas as pd
# Convert to DSPy Example objects
dspy_examples = {}
for section_name, section in dataset.items():
    dspy_examples[section_name] = [
        dspy.Example(
            premise=ex['premise'],
            hypothesis=ex['hypothesis'],
            label=label_names[ex['gold_label']]
        ).with_inputs("premise", "hypothesis")
        for ex in section
    ]

df = pd.DataFrame(dspy_examples)

presupposition_all_n_presupposition presupposition_both_presupposition  \
0           [premise, hypothesis, label]       [premise, hypothesis, label]   
1           [premise, hypothesis, label]       [premise, hypothesis, label]   
2           [premise, hypothesis, label]       [premise, hypothesis, label]   
3           [premise, hypothesis, label]       [premise, hypothesis, label]   
4           [premise, hypothesis, label]       [premise, hypothesis, label]   
...                                  ...                                ...   
1895        [premise, hypothesis, label]       [premise, hypothesis, label]   
1896        [premise, hypothesis, label]       [premise, hypothesis, label]   
1897        [premise, hypothesis, label]       [premise, hypothesis, label]   
1898        [premise, hypothesis, label]       [premise, hypothesis, label]   
1899        [premise, hypothesis, label]       [premise, hypothesis, label]   

     presupposition_change_of_state presupposition_cleft_existence  \
0      [premise, hypothesis, label]   [premise, hypothesis, label]   
1      [premise, hypothesis, label]   [premise, hypothesis, label]   
2      [premise, hypothesis, label]   [premise, hypothesis, label]   
3      [premise, hypothesis, label]   [premise, hypothesis, label]   
4      [premise, hypothesis, label]   [premise, hypothesis, label]   
...                             ...                            ...   
1895   [premise, hypothesis, label]   [premise, hypothesis, label]   
1896   [premise, hypothesis, label]   [premise, hypothesis, label]   
1897   [premise, hypothesis, label]   [premise, hypothesis, label]   
1898   [premise, hypothesis, label]   [premise, hypothesis, label]   
1899   [premise, hypothesis, label]   [premise, hypothesis, label]   

     presupposition_cleft_uniqueness presupposition_only_presupposition  \
0       [premise, hypothesis, label]       [premise, hypothesis, label]   
1       [premise, hypothesis, label]       [premise, hypothesis, label]   
2       [premise, hypothesis, label]       [premise, hypothesis, label]   
3       [premise, hypothesis, label]       [premise, hypothesis, label]   
4       [premise, hypothesis, label]       [premise, hypothesis, label]   
...                              ...                                ...   
1895    [premise, hypothesis, label]       [premise, hypothesis, label]   
1896    [premise, hypothesis, label]       [premise, hypothesis, label]   
1897    [premise, hypothesis, label]       [premise, hypothesis, label]   
1898    [premise, hypothesis, label]       [premise, hypothesis, label]   
1899    [premise, hypothesis, label]       [premise, hypothesis, label]   

     presupposition_possessed_definites_existence  \
0                    [premise, hypothesis, label]   
1                    [premise, hypothesis, label]   
2                    [premise, hypothesis, label]   
3                    [premise, hypothesis, label]   
4                    [premise, hypothesis, label]   
...                                           ...   
1895                 [premise, hypothesis, label]   
1896                 [premise, hypothesis, label]   
1897                 [premise, hypothesis, label]   
1898                 [premise, hypothesis, label]   
1899                 [premise, hypothesis, label]   

     presupposition_possessed_definites_uniqueness  \
0                     [premise, hypothesis, label]   
1                     [premise, hypothesis, label]   
2                     [premise, hypothesis, label]   
3                     [premise, hypothesis, label]   
4                     [premise, hypothesis, label]   
...                                            ...   
1895                  [premise, hypothesis, label]   
1896                  [premise, hypothesis, label]   
1897                  [premise, hypothesis, label]   
1898                  [premise, hypothesis, label]   
1899                  [premise, hypothesis, label]   

     presupposition_question

In [ ]:
from dspy.evaluate import Evaluate
from evaluate import combine, load

# 1. Run DSPy evaluation for each section (here, limited to first 10 for demo)
results = {}  # Store per-section predictions
not_predicted = {}
for sec in dspy_examples:
    print(f"Evaluating section:\t{sec}")
    evaluator = Evaluate(
        devset=dspy_examples[sec],
        metric=accuracy_metric,
        return_outputs=True,
        num_threads=50,
        display_progress=True,
        display_table=False,
        provide_traceback=False
        # max_errors=30
    )
    eval_res = evaluator(predictor)
    _, result_tuples = eval_res
    print(f"number of results:\t{len(result_tuples)}")
    preds, refs = [], []
    not_predicted[sec] = {
        'section':sec,
        'num_not_predicted':0,
        'not_predicted':[]
    }
    for example, prediction, correct in result_tuples:
        if not hasattr(prediction, "label"):
            not_predicted[sec]['num_not_predicted']+=1
            not_predicted[sec]['not_predicted'].append((example, prediction, correct))
            continue
        preds.append(prediction.label)
        refs.append(example.label)
    results[sec] = {"preds": preds, "refs": refs}

Let's display some statistics about the results

In [ ]:
from collections import Counter

for sec, data in results.items():
    preds = data['preds']
    refs = data['refs']
    print(f"Section: {sec}")
    print(f"  Total predictions: {len(preds)}")
    print(f"  Total references:  {len(refs)}")
    print(f"  Class distribution in predictions: {Counter(preds)}")
    print(f"  Class distribution in references:  {Counter(refs)}")
    agree = sum([p == r for p, r in zip(preds, refs)])
    print(f"  Number of matches (agreement): {agree}")
    print(f"  Accuracy (quick): {agree / len(refs):.3f}")
    print()

# Overall stats
all_preds = sum([v['preds'] for v in results.values()], [])
all_refs  = sum([v['refs']  for v in results.values()], [])
print("=== OVERALL ===")
print(f"Total predictions: {len(all_preds)}")
print(f"Total references:  {len(all_refs)}")
print(f"Class distribution in predictions: {Counter(all_preds)}")
print(f"Class distribution in references:  {Counter(all_refs)}")
agree = sum([p == r for p, r in zip(all_preds, all_refs)])
print(f"Number of matches (agreement): {agree}")
print(f"Accuracy (quick): {agree / len(all_refs):.3f}")


We will now show information about non-predicted examples:

In [ ]:
df_np = pd.DataFrame(list(not_predicted.values())).set_index("section")
exploded = df_np["not_predicted"].explode()
df_details = (
    exploded
    .reset_index()
    .rename(columns={"index": "section", "not_predicted": "detail"})
    .join(pd.json_normalize(exploded).add_prefix("detail."))
)
display(df_details)
for sec, info in not_predicted.items():
    print(f"=== Section: {sec} — {info['num_not_predicted']} failures ===")
    for ex, raw_out, score in info['not_predicted']:
        print(ex)
        premise, hypothesis, ref,= ex
        print(f"🎯 Ref label: {ex[ref]}")
        print(f"💬 Premise: {ex[premise]}")
        print(f"💬 Hypothesis: {ex[hypothesis]}")
        print(f"🛑 Raw output: {raw_out!r}")
        print(f"⚠️ Score: {score}")
        print("-" * 40)

In [ ]:
# 2. Prepare for metric calculation
metric_prf = combine(["precision", "recall", "f1"])
acc = load("accuracy")
rows = []
all_preds, all_refs = [], []
label2id = {"entailment": 0, "neutral": 1, "contradiction": 2}

for sec, data in results.items():
    print(f"Computing metrics for section: {sec}")
    preds = [label2id[label] for label in data["preds"]]
    refs  = [label2id[label] for label in data["refs"]]
    prf = metric_prf.compute(predictions=preds, references=refs, average="weighted")
    accuracy = acc.compute(predictions=preds, references=refs)["accuracy"]

    rows.append({"section": sec, "accuracy": accuracy, **prf})
    all_preds += preds
    all_refs += refs

# 3. Compute overall metrics
overall_prf = metric_prf.compute(predictions=all_preds, references=all_refs, average="weighted")
overall_acc = acc.compute(predictions=all_preds, references=all_refs)["accuracy"]
rows.append({"section": "all", "accuracy": overall_acc, **overall_prf})

# Create DataFrame and display
df_metrics = pd.DataFrame(rows)
display(df_metrics.set_index("section"))

In our experiment we got the following results:
| section                                       | accuracy | precision | recall  | f1       |
|----------------------------------------------|----------|-----------|---------|----------|
| presupposition_all_n_presupposition          | 0.942632 | 0.949257  | 0.942632| 0.942783 |
| presupposition_both_presupposition           | 0.973158 | 0.974034  | 0.973158| 0.973184 |
| presupposition_change_of_state               | 0.557895 | 0.655905  | 0.557895| 0.493381 |
| presupposition_cleft_existence               | 0.686316 | 0.812531  | 0.686316| 0.669707 |
| presupposition_cleft_uniqueness              | 0.474211 | 0.503028  | 0.474211| 0.350207 |
| presupposition_only_presupposition           | 0.668947 | 0.778061  | 0.668947| 0.654415 |
| presupposition_possessed_definites_existence | 0.923158 | 0.929153  | 0.923158| 0.923322 |
| presupposition_possessed_definites_uniqueness| 0.475263 | 0.626211  | 0.475263| 0.352235 |
| presupposition_question_presupposition       | 0.841053 | 0.863356  | 0.841053| 0.838288 |
| all                                          | 0.726959 | 0.815532  | 0.726959| 0.717863 |

With a total F1 score of 0.726959 with grok-3-mini. Let's try to optimize the model


## Optimizing the model
We are going to try optimize the model in a couple ways.
we will first create a dev\test split:

In [57]:
# acc_metric   = evaluate.load("accuracy")
# f1_metric    = evaluate.load("f1")
# prec_metric  = evaluate.load("precision")
# rec_metric   = evaluate.load("recall")

In [47]:
from numpy import random
rng = random.default_rng(42)

def stratified_split(df, frac=0.7):
    idx_dev = []
    for (sec, lab), g in df.groupby(["section","gold_label"]):
        n = int(len(g)*frac)
        idx = rng.permutation(g.index)
        idx_dev.extend(idx[:n])
    dev = df.loc[idx_dev]
    test = df.drop(idx_dev)
    return dev, test

dev_df, test_df = stratified_split(combined_df)

In [48]:
display(pd.DataFrame(dev_df))
display(pd.DataFrame(test_df))

premise  \
1795   All eight women that compel libraries to appre...   
500    Have the six guests that had badgered a lot of...   
1792   Do all eight women that compel libraries to ap...   
174    The ten women that talk didn't implore Pamela ...   
152    All four waiters that were boring Paul have te...   
...                                                  ...   
15956  Patricia wasn't concealing when Renee commande...   
15324  Museums might know when Donald neglects to hid...   
15961      Sonia does forget when those peppers blacken.   
16328        Does Jill conceal why the girls do scratch?   
16920  Marla might find out why these waitresses have...   

                                              hypothesis         trigger  \
1795   There are exactly eight women that compel libr...           modal   
500    There are exactly six guests that had badgered...   interrogative   
1792   There are exactly eight women that compel libr...   interrogative   
174               There are exactly ten women that talk.         negated   
152    There are exactly four waiters that were borin...      unembedded   
...                                                  ...             ...   
15956  Patricia is concealing when Renee commanded th...  Not_In_Example   
15324   Donald doesn't neglect to hide that wheelbarrow.           modal   
15961                       Those peppers don't blacken.      unembedded   
16328                           The girls don't scratch.   interrogative   
16920               These waitresses haven't retaliated.           modal   

             trigger1        trigger2  presupposition  gold_label  \
1795   Not_In_Example  Not_In_Example        positive           0   
500    Not_In_Example  Not_In_Example        positive           0   
1792   Not_In_Example  Not_In_Example        positive           0   
174    Not_In_Example  Not_In_Example        positive           0   
152    Not_In_Example  Not_In_Example        positive           0   
...               ...             ...             ...         ...   
15956         negated      unembedded  Not_In_Example           2   
15324  Not_In_Example  Not_In_Example         negated           2   
15961  Not_In_Example  Not_In_Example         negated           2   
16328  Not_In_Example  Not_In_Example         negated           2   
16920  Not_In_Example  Not_In_Example         negated           2   

                           UID pairID  paradigmID  \
1795      all_n_presupposition  1795e          94   
500       all_n_presupposition   500e          26   
1792      all_n_presupposition  1792e          94   
174       all_n_presupposition   174e           9   
152       all_n_presupposition   152e           8   
...                        ...    ...         ...   
15956  question_presupposition   756c          39   
15324  question_presupposition   124c           6   
15961  question_presupposition   761c          40   
16328  question_presupposition  1128c          59   
16920  question_presupposition  1720c          90   

                                      section  
1795      presupposition_all_n_presupposition  
500       presupposition_all_n_presupposition  
1792      presupposition_all_n_presupposition  
174       presupposition_all_n_presupposition  
152       presupposition_all_n_presupposition  
...                                       ...  
15956  presupposition_question_presupposition  
15324  presupposition_question_presupposition  
15961  presupposition_question_presupposition  
16328  presupposition_question_presupposition  
16920  presupposition_question_presupposition  

[11970 rows x 11 columns]

premise  \
3      All ten guys that proved to boast weren't divo...   
4      All ten guys that proved to boast weren't divo...   
10     All ten guys that proved to boast might have b...   
14     If all ten guys that proved to boast were divo...   
19     All ten reports that can bore some waiter are ...   
...                                                  ...   
17087  Did the actors conceal where that mall shocks ...   
17090  The actors might have concealed where that mal...   
17092  The actors might have concealed where that mal...   
17093  If the actors do conceal where that mall shock...   
17096  The actors didn't conceal where that mall shoc...   

                                              hypothesis         trigger  \
3       There are exactly ten guys that proved to boast.         negated   
4      There are exactly eleven guys that proved to b...         negated   
10     There are exactly eleven guys that proved to b...           modal   
14     There are exactly ten senators that proved to ...     conditional   
19     There are exactly ten reports that can bore so...      unembedded   
...                                                  ...             ...   
17087                             That mall shocks Janet   interrogative   
17090                             That mall shocks Janet           modal   
17092                               Travel shocks Janet.           modal   
17093                             That mall shocks Janet     conditional   
17096  The actors do conceal where that mall shocks J...  Not_In_Example   

             trigger1        trigger2  presupposition  gold_label  \
3      Not_In_Example  Not_In_Example        positive           0   
4      Not_In_Example  Not_In_Example         negated           2   
10     Not_In_Example  Not_In_Example         negated           2   
14     Not_In_Example  Not_In_Example         neutral           1   
19     Not_In_Example  Not_In_Example        positive           0   
...               ...             ...             ...         ...   
17087  Not_In_Example  Not_In_Example        positive           0   
17090  Not_In_Example  Not_In_Example        positive           0   
17092  Not_In_Example  Not_In_Example         neutral           1   
17093  Not_In_Example  Not_In_Example        positive           0   
17096         negated      unembedded  Not_In_Example           2   

                           UID pairID  paradigmID  \
3         all_n_presupposition     3e           0   
4         all_n_presupposition     4c           0   
10        all_n_presupposition    10c           0   
14        all_n_presupposition    14n           0   
19        all_n_presupposition    19e           1   
...                        ...    ...         ...   
17087  question_presupposition  1887e          99   
17090  question_presupposition  1890e          99   
17092  question_presupposition  1892n          99   
17093  question_presupposition  1893e          99   
17096  question_presupposition  1896c          99   

                                      section  
3         presupposition_all_n_presupposition  
4         presupposition_all_n_presupposition  
10        presupposition_all_n_presupposition  
14        presupposition_all_n_presupposition  
19        presupposition_all_n_presupposition  
...                                       ...  
17087  presupposition_question_presupposition  
17090  presupposition_question_presupposition  
17092  presupposition_question_presupposition  
17093  presupposition_question_presupposition  
17096  presupposition_question_presupposition  

[5130 rows x 11 columns]

In [49]:
def to_examples(df):
    return [dspy.Example(
        premise=r.premise, hypothesis=r.hypothesis,
        label=label_names[r.gold_label]
    ).with_inputs("premise","hypothesis") for r in df.itertuples()]
dev_ex  = to_examples(dev_df)
test_ex = to_examples(test_df)

In [ ]:
def evaluate(model):
    results = Evaluate(
        devset=test_ex,
        metric=accuracy_metric,
        return_outputs=True,
        num_threads=50,
        display_progress=True,
        display_table=False,
        provide_traceback=False
    )(model)
    score,results = results
    print(f"Score:\t{score}")
    test_pred = [label2id[out[1].label] for out in results]
    return score, results, test_pred
def compute_matrices(test_pred):
    prf = metric_prf.compute(predictions=test_pred, references=y_true, average="weighted")
    accuracy = acc.compute(predictions=test_pred, references=y_true)
    return {**prf, **accuracy}

In [58]:
predictor_test_predictions = Evaluate(
        devset=test_ex,
        metric=accuracy_metric,
        return_outputs=True,
        num_threads=50,
        display_progress=True,
        display_table=False,
        provide_traceback=False
)(predictor)

Average Metric: 2278.00 / 2977 (76.5%):  58%|█████▊    | 2976/5130 [00:02<00:00, 2598.10it/s]

IOPub message rate exceeded.
The Jupyter server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--ServerApp.iopub_msg_rate_limit`.

Current values:
ServerApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
ServerApp.rate_limit_window=3.0 (secs)



In [101]:
score,predictor_test_predictions_results = predictor_test_predictions
print(f"Score: {score}")
predictor_test_pred = [label2id[out[1].label] for out in predictor_test_predictions_results]
y_true = [label2id[ex.label]  for ex in test_ex]

Score: 72.48


In [102]:
predictor_prf = metric_prf.compute(predictions=predictor_test_pred, references=y_true, average="weighted")
predictor_accuracy = acc.compute(predictions=predictor_test_pred, references=y_true)
predictor_combined = {**predictor_prf, **predictor_accuracy}

In [103]:
display(pd.DataFrame(predictor_combined, index=["Original predictor"]))

precision    recall        f1  accuracy
Original predictor   0.813633  0.724756  0.715039  0.724756

In our code we saw a F1 score of 0.715039 on the test.

### Simply few-shot strategy over the entire dataset

In [50]:
from dspy.teleprompt import BootstrapFewShot
bs = BootstrapFewShot(metric=accuracy_metric, max_bootstrapped_demos=20, max_labeled_demos=16)
optimized = bs.compile(student=predictor, trainset=dev_ex)

  0%|          | 23/11970 [02:15<19:36:13,  5.91s/it]

Bootstrapped 20 full traces after 23 examples for up to 1 rounds, amounting to 23 attempts.


In [52]:
overall_optimized = optimized

In [55]:
# 3. Evaluate
overall_report = Evaluate(
        devset=test_ex,
        metric=accuracy_metric,
        return_outputs=True,
        num_threads=50,
        display_progress=True,
        display_table=False,
        provide_traceback=False
)(overall_optimized)

Average Metric: 3914.00 / 5130 (76.3%): 100%|██████████| 5130/5130 [10:32<00:00,  8.12it/s]  

2025/07/23 17:46:20 INFO dspy.evaluate.evaluate: Average Metric: 3914 / 5130 (76.3%)


In [104]:
# 2) Extract labels
overall_score,overall_report_results = overall_report
print(f"Score: {overall_score}")
overall_test_pred = [label2id[out[1].label] for out in overall_report_results]

Score: 76.3


In [105]:
overall_prf = metric_prf.compute(predictions=overall_test_pred, references=y_true, average="weighted")
overall_accuracy = acc.compute(predictions=overall_test_pred, references=y_true)
overall_combined = {**overall_prf, **overall_accuracy}

In [106]:
display(pd.DataFrame(overall_combined, index=["Overall_optimized predictor"]))

precision    recall        f1  accuracy
Overall_optimized predictor   0.826545  0.762963  0.758567  0.762963

When testing the overall model, we saw F1 Score of 0.758567, an improvement of 6.087%!

### Adaptive few-shot strategy
We will now try to optimize for each section and create a new model which will predicate by majority vote.

In [110]:
sec_dev_ex = {sec: to_examples(group) for sec, group in dev_df.groupby("section")}

In [111]:
optimized_pipelines = {}

for sec in sec_dev_ex:
    print(f"Optimizing for section: {sec}")
    # Flatten dev examples for prompt tuning
    dev_set = sec_dev_ex[sec]

    # Initialize optimizer
    bs = BootstrapFewShot(
        metric=accuracy_metric,
        max_bootstrapped_demos=8,
        max_labeled_demos=4
    )

    # Compile and tune using dev split
    compiled = bs.compile(
        student=predictor,
        trainset=dev_set
    )
    optimized_pipelines[sec] = compiled

Optimizing for section: presupposition_all_n_presupposition
Average Metric: 94.00 / 94 (100.0%):   1%|          | 94/11970 [54:37<115:01:59, 34.87s/it]


  1%|          | 8/1330 [00:59<2:43:43,  7.43s/it]


Bootstrapped 8 full traces after 8 examples for up to 1 rounds, amounting to 8 attempts.
✅ Completed Bootstrapped few-shot for section `presupposition_all_n_presupposition`
Optimizing for section: presupposition_both_presupposition


  1%|          | 8/1330 [00:53<2:26:49,  6.66s/it]


Bootstrapped 8 full traces after 8 examples for up to 1 rounds, amounting to 8 attempts.
✅ Completed Bootstrapped few-shot for section `presupposition_both_presupposition`
Optimizing for section: presupposition_change_of_state


  2%|▏         | 27/1330 [02:21<1:54:10,  5.26s/it]


Bootstrapped 8 full traces after 27 examples for up to 1 rounds, amounting to 27 attempts.
✅ Completed Bootstrapped few-shot for section `presupposition_change_of_state`
Optimizing for section: presupposition_cleft_existence


  2%|▏         | 21/1330 [02:02<2:07:14,  5.83s/it]


Bootstrapped 8 full traces after 21 examples for up to 1 rounds, amounting to 21 attempts.
✅ Completed Bootstrapped few-shot for section `presupposition_cleft_existence`
Optimizing for section: presupposition_cleft_uniqueness


  3%|▎         | 36/1330 [04:14<2:32:31,  7.07s/it]


Bootstrapped 8 full traces after 36 examples for up to 1 rounds, amounting to 36 attempts.
✅ Completed Bootstrapped few-shot for section `presupposition_cleft_uniqueness`
Optimizing for section: presupposition_only_presupposition


  2%|▏         | 20/1330 [01:46<1:56:17,  5.33s/it]


Bootstrapped 8 full traces after 20 examples for up to 1 rounds, amounting to 20 attempts.
✅ Completed Bootstrapped few-shot for section `presupposition_only_presupposition`
Optimizing for section: presupposition_possessed_definites_existence


  1%|          | 8/1330 [00:37<1:42:59,  4.67s/it]


Bootstrapped 8 full traces after 8 examples for up to 1 rounds, amounting to 8 attempts.
✅ Completed Bootstrapped few-shot for section `presupposition_possessed_definites_existence`
Optimizing for section: presupposition_possessed_definites_uniqueness


 14%|█▍        | 183/1330 [19:21<2:01:20,  6.35s/it]


Bootstrapped 8 full traces after 183 examples for up to 1 rounds, amounting to 183 attempts.
✅ Completed Bootstrapped few-shot for section `presupposition_possessed_definites_uniqueness`
Optimizing for section: presupposition_question_presupposition


  1%|          | 14/1330 [01:23<2:10:47,  5.96s/it]

Bootstrapped 8 full traces after 14 examples for up to 1 rounds, amounting to 14 attempts.
✅ Completed Bootstrapped few-shot for section `presupposition_question_presupposition`


In [112]:
# existing section pipelines
ensemble_tp = dspy.Ensemble(reduce_fn=dspy.majority)
adaptive_optimized = ensemble_tp.compile(list(optimized_pipelines.values()))
#
# # 3. Now save the compiled predictor to disk
# fname = f"joint_predictor_state_{datetime.now().timestamp():.0f}.pkl"
# joint_predictor.save(fname, save_program=False)
# print(f"✅ Saved joint predictor to {fname}")

In [113]:
adaptive_report = Evaluate(
        devset=test_ex,
        metric=accuracy_metric,
        return_outputs=True,
        num_threads=50,
        display_progress=True,
        display_table=False,
        provide_traceback=False
)(adaptive_optimized)

Average Metric: 3821.00 / 5130 (74.5%): 100%|██████████| 5130/5130 [1:37:27<00:00,  1.14s/it]  

2025/07/23 20:42:31 INFO dspy.evaluate.evaluate: Average Metric: 3821 / 5130 (74.5%)


In [117]:
adaptive_score,adaptive_report_results = adaptive_report
print(f"Score: {adaptive_score}")
adaptive_test_pred = [label2id[out[1].label] for out in adaptive_report_results]

Score: 74.48


In [118]:
adaptive_prf = metric_prf.compute(predictions=adaptive_test_pred, references=y_true, average="weighted")
adaptive_accuracy = acc.compute(predictions=adaptive_test_pred, references=y_true)
adaptive_combined = {**adaptive_prf, **adaptive_accuracy}

In [119]:
display(pd.DataFrame(adaptive_combined, index=["Adaptive_optimized predictor"]))

precision    recall        f1  accuracy
Adaptive_optimized predictor   0.821135  0.744834  0.738229  0.744834

We got:
0.821135,0.744834,0.738229,0.744834
this shows around 0.02 improvement.

In [120]:
from dspy.teleprompt import BootstrapFewShotWithRandomSearch

bsrs = BootstrapFewShotWithRandomSearch(
    metric=accuracy_metric,
    max_bootstrapped_demos=20,
    max_labeled_demos=9,
    num_candidate_programs=5,
    num_threads=50
)
opted_rs = bsrs.compile(student=predictor, trainset=dev_ex, valset=dev_ex)

Going to sample between 1 and 20 traces per predictor.
Will attempt to bootstrap 5 candidate sets.
Average Metric: 2500.00 / 2985 (83.8%):  25%|██▍       | 2984/11970 [00:05<00:04, 1873.90it/s]

IOPub message rate exceeded.
The Jupyter server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--ServerApp.iopub_msg_rate_limit`.

Current values:
ServerApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
ServerApp.rate_limit_window=3.0 (secs)



Average Metric: 7203.00 / 9896 (72.8%):  83%|████████▎ | 9895/11970 [00:08<00:01, 1622.00it/s]

IOPub message rate exceeded.
The Jupyter server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--ServerApp.iopub_msg_rate_limit`.

Current values:
ServerApp.iopub_msg_rate_limit=1000.0 (msgs/sec)
ServerApp.rate_limit_window=3.0 (secs)



Average Metric: 242.00 / 255 (94.9%):   2%|▏         | 254/11970 [00:53<3:04:36,  1.06it/s]

IOStream.flush timed out


Average Metric: 9258.00 / 11970 (77.3%): 100%|██████████| 11970/11970 [24:56<00:00,  8.00it/s]


2025/07/23 21:17:57 INFO dspy.evaluate.evaluate: Average Metric: 9258 / 11970 (77.3%)


New best score: 77.34 for seed -2
Scores so far: [72.81, 77.34]
Best score so far: 77.34


  0%|          | 20/11970 [00:00<01:55, 103.60it/s]


Bootstrapped 20 full traces after 20 examples for up to 1 rounds, amounting to 20 attempts.
Average Metric: 9169.00 / 11970 (76.6%): 100%|██████████| 11970/11970 [24:24<00:00,  8.17it/s]

2025/07/23 21:42:56 INFO dspy.evaluate.evaluate: Average Metric: 9169 / 11970 (76.6%)



Scores so far: [72.81, 77.34, 76.6]
Best score so far: 77.34


  0%|          | 18/11970 [01:11<13:15:31,  3.99s/it]


Bootstrapped 13 full traces after 18 examples for up to 1 rounds, amounting to 18 attempts.
Average Metric: 7164.00 / 9521 (75.2%):  80%|███████▉  | 9520/11970 [19:38<1:32:35,  2.27s/it]

IOStream.flush timed out


Average Metric: 7555.00 / 10048 (75.2%):  84%|████████▍ | 10047/11970 [20:19<02:00, 15.93it/s]

2025/07/23 22:05:11 ERROR dspy.utils.parallelizer: Error for Example({'premise': "If Irene's print that hadn't bothered Cindy does fall, it's okay.", 'hypothesis': "Irene's print that hadn't bothered Cindy does fall.", 'label': 'neutral'}) (input_keys={'hypothesis', 'premise'}): Both structured output format and JSON mode failed. Please choose a model that supports `response_format` argument. Original error: litellm.RateLimitError: RateLimitError: XaiException - Error code: 429 - {'code': 'Some resource has been exhausted', 'error': "Too many requests for team bc29b82a-19d1-4c03-acf7-cd8b768453be and model grok-3-mini. Your team's rate limit is — Requests per Second (actual/limit): 0/8, Requests per Minute (actual/limit): 495/480. You can view your team rate limits at https://console.x.ai."}. Set `provide_traceback=True` for traceback.


Average Metric: 7555.00 / 10048 (75.2%):  84%|████████▍ | 10049/11970 [20:36<1:09:21,  2.17s/it]

2025/07/23 22:05:11 ERROR dspy.utils.parallelizer: Error for Example({'premise': "Did Mitchell's mountain that is hurting Marcus disgust Timothy?", 'hypothesis': "Mitchell's mountain that is hurting Marcus did disgust Timothy.", 'label': 'neutral'}) (input_keys={'hypothesis', 'premise'}): Both structured output format and JSON mode failed. Please choose a model that supports `response_format` argument. Original error: litellm.RateLimitError: RateLimitError: XaiException - Error code: 429 - {'code': 'Some resource has been exhausted', 'error': "Too many requests for team bc29b82a-19d1-4c03-acf7-cd8b768453be and model grok-3-mini. Your team's rate limit is — Requests per Second (actual/limit): 0/8, Requests per Minute (actual/limit): 495/480. You can view your team rate limits at https://console.x.ai."}. Set `provide_traceback=True` for traceback.
2025/07/23 22:05:11 ERROR dspy.utils.parallelizer: Error for Example({'premise': "If Theresa's river that hasn't hurt Adam did freeze, it's ok

Average Metric: 7555.00 / 10048 (75.2%):  84%|████████▍ | 10050/11970 [20:36<59:44,  1.87s/it]  

2025/07/23 22:05:11 ERROR dspy.utils.parallelizer: Error for Example({'premise': "Homer's snake that embarrassed some legislature did resemble the paintings.", 'hypothesis': 'Joel has exactly one snake that embarrassed some legislature.', 'label': 'neutral'}) (input_keys={'hypothesis', 'premise'}): Both structured output format and JSON mode failed. Please choose a model that supports `response_format` argument. Original error: litellm.RateLimitError: RateLimitError: XaiException - Error code: 429 - {'code': 'Some resource has been exhausted', 'error': "Too many requests for team bc29b82a-19d1-4c03-acf7-cd8b768453be and model grok-3-mini. Your team's rate limit is — Requests per Second (actual/limit): 0/8, Requests per Minute (actual/limit): 495/480. You can view your team rate limits at https://console.x.ai."}. Set `provide_traceback=True` for traceback.


Average Metric: 7555.00 / 10048 (75.2%):  84%|████████▍ | 10052/11970 [20:36<42:36,  1.33s/it]

2025/07/23 22:05:11 ERROR dspy.utils.parallelizer: Error for Example({'premise': "Does Wendy's cup that hasn't vanished disturb Nicole?", 'hypothesis': "Wendy's cup that hasn't vanished does disturb Nicole.", 'label': 'neutral'}) (input_keys={'hypothesis', 'premise'}): Both structured output format and JSON mode failed. Please choose a model that supports `response_format` argument. Original error: litellm.RateLimitError: RateLimitError: XaiException - Error code: 429 - {'code': 'Some resource has been exhausted', 'error': "Too many requests for team bc29b82a-19d1-4c03-acf7-cd8b768453be and model grok-3-mini. Your team's rate limit is — Requests per Second (actual/limit): 0/8, Requests per Minute (actual/limit): 495/480. You can view your team rate limits at https://console.x.ai."}. Set `provide_traceback=True` for traceback.


Average Metric: 7555.00 / 10048 (75.2%):  84%|████████▍ | 10052/11970 [20:36<42:36,  1.33s/it]

2025/07/23 22:05:12 ERROR dspy.utils.parallelizer: Error for Example({'premise': "Sally's car that annoys Vincent might bore Tina.", 'hypothesis': "Sally's car that annoys Vincent does bore Tina.", 'label': 'neutral'}) (input_keys={'hypothesis', 'premise'}): Both structured output format and JSON mode failed. Please choose a model that supports `response_format` argument. Original error: litellm.RateLimitError: RateLimitError: XaiException - Error code: 429 - {'code': 'Some resource has been exhausted', 'error': "Too many requests for team bc29b82a-19d1-4c03-acf7-cd8b768453be and model grok-3-mini. Your team's rate limit is — Requests per Second (actual/limit): 0/8, Requests per Minute (actual/limit): 493/480. You can view your team rate limits at https://console.x.ai."}. Set `provide_traceback=True` for traceback.


Average Metric: 7555.00 / 10048 (75.2%):  84%|████████▍ | 10054/11970 [20:36<32:53,  1.03s/it]

2025/07/23 22:05:12 ERROR dspy.utils.parallelizer: Error for Example({'premise': "Did Becca's horse that existed break the plane?", 'hypothesis': "Becca's horse that existed did break the plane.", 'label': 'neutral'}) (input_keys={'hypothesis', 'premise'}): Both structured output format and JSON mode failed. Please choose a model that supports `response_format` argument. Original error: litellm.RateLimitError: RateLimitError: XaiException - Error code: 429 - {'code': 'Some resource has been exhausted', 'error': "Too many requests for team bc29b82a-19d1-4c03-acf7-cd8b768453be and model grok-3-mini. Your team's rate limit is — Requests per Second (actual/limit): 0/8, Requests per Minute (actual/limit): 493/480. You can view your team rate limits at https://console.x.ai."}. Set `provide_traceback=True` for traceback.


Average Metric: 7555.00 / 10048 (75.2%):  84%|████████▍ | 10054/11970 [20:36<32:53,  1.03s/it]

2025/07/23 22:05:12 ERROR dspy.utils.parallelizer: Error for Example({'premise': "Karla's snake that had upset Vanessa did bore Veronica.", 'hypothesis': 'The customer has exactly one snake that had upset Vanessa.', 'label': 'neutral'}) (input_keys={'hypothesis', 'premise'}): Both structured output format and JSON mode failed. Please choose a model that supports `response_format` argument. Original error: litellm.RateLimitError: RateLimitError: XaiException - Error code: 429 - {'code': 'Some resource has been exhausted', 'error': "Too many requests for team bc29b82a-19d1-4c03-acf7-cd8b768453be and model grok-3-mini. Your team's rate limit is — Requests per Second (actual/limit): 0/8, Requests per Minute (actual/limit): 493/480. You can view your team rate limits at https://console.x.ai."}. Set `provide_traceback=True` for traceback.


Average Metric: 7555.00 / 10048 (75.2%):  84%|████████▍ | 10055/11970 [20:37<32:52,  1.03s/it]

2025/07/23 22:05:12 ERROR dspy.utils.parallelizer: Error for Example({'premise': "Did Theresa's river that hasn't hurt Adam freeze?", 'hypothesis': "Waitresses has exactly one river that hasn't hurt Adam.", 'label': 'neutral'}) (input_keys={'hypothesis', 'premise'}): Both structured output format and JSON mode failed. Please choose a model that supports `response_format` argument. Original error: litellm.RateLimitError: RateLimitError: XaiException - Error code: 429 - {'code': 'Some resource has been exhausted', 'error': "Too many requests for team bc29b82a-19d1-4c03-acf7-cd8b768453be and model grok-3-mini. Your team's rate limit is — Requests per Second (actual/limit): 0/8, Requests per Minute (actual/limit): 494/480. You can view your team rate limits at https://console.x.ai."}. Set `provide_traceback=True` for traceback.


Average Metric: 7555.00 / 10048 (75.2%):  84%|████████▍ | 10057/11970 [20:37<20:49,  1.53it/s]

2025/07/23 22:05:12 ERROR dspy.utils.parallelizer: Error for Example({'premise': "Does those dancers' closet that aggravates many women resemble those photographs?", 'hypothesis': "Those dancers' closet that aggravates many women does resemble those photographs.", 'label': 'neutral'}) (input_keys={'hypothesis', 'premise'}): Both structured output format and JSON mode failed. Please choose a model that supports `response_format` argument. Original error: litellm.RateLimitError: RateLimitError: XaiException - Error code: 429 - {'code': 'Some resource has been exhausted', 'error': "Too many requests for team bc29b82a-19d1-4c03-acf7-cd8b768453be and model grok-3-mini. Your team's rate limit is — Requests per Second (actual/limit): 0/8, Requests per Minute (actual/limit): 494/480. You can view your team rate limits at https://console.x.ai."}. Set `provide_traceback=True` for traceback.


Average Metric: 7555.00 / 10048 (75.2%):  84%|████████▍ | 10057/11970 [20:37<20:49,  1.53it/s]

2025/07/23 22:05:12 ERROR dspy.utils.parallelizer: Error for Example({'premise': "Brian's popsicle that couldn't stun Karla might fall.", 'hypothesis': "Brian's popsicle that couldn't stun Karla does fall.", 'label': 'neutral'}) (input_keys={'hypothesis', 'premise'}): Both structured output format and JSON mode failed. Please choose a model that supports `response_format` argument. Original error: litellm.RateLimitError: RateLimitError: XaiException - Error code: 429 - {'code': 'Some resource has been exhausted', 'error': "Too many requests for team bc29b82a-19d1-4c03-acf7-cd8b768453be and model grok-3-mini. Your team's rate limit is — Requests per Second (actual/limit): 0/8, Requests per Minute (actual/limit): 493/480. You can view your team rate limits at https://console.x.ai."}. Set `provide_traceback=True` for traceback.


Average Metric: 7555.00 / 10048 (75.2%):  84%|████████▍ | 10059/11970 [20:37<20:47,  1.53it/s]

2025/07/23 22:05:12 ERROR dspy.utils.parallelizer: Error for Example({'premise': "Does Elizabeth's couch that wasn't stunning Rhonda slip?", 'hypothesis': "Elizabeth's couch that wasn't stunning Rhonda does slip.", 'label': 'neutral'}) (input_keys={'hypothesis', 'premise'}): Both structured output format and JSON mode failed. Please choose a model that supports `response_format` argument. Original error: litellm.RateLimitError: RateLimitError: XaiException - Error code: 429 - {'code': 'Some resource has been exhausted', 'error': "Too many requests for team bc29b82a-19d1-4c03-acf7-cd8b768453be and model grok-3-mini. Your team's rate limit is — Requests per Second (actual/limit): 0/8, Requests per Minute (actual/limit): 494/480. You can view your team rate limits at https://console.x.ai."}. Set `provide_traceback=True` for traceback.


Average Metric: 7555.00 / 10048 (75.2%):  85%|████████▌ | 10231/11970 [20:37<00:47, 36.58it/s]

2025/07/23 22:05:12 ERROR dspy.utils.parallelizer: Error for Example({'premise': "If Becca's horse that existed did break the plane, it's okay.", 'hypothesis': "Becca's horse that existed did break the plane.", 'label': 'neutral'}) (input_keys={'hypothesis', 'premise'}): Both structured output format and JSON mode failed. Please choose a model that supports `response_format` argument. Original error: litellm.RateLimitError: RateLimitError: XaiException - Error code: 429 - {'code': 'Some resource has been exhausted', 'error': "Too many requests for team bc29b82a-19d1-4c03-acf7-cd8b768453be and model grok-3-mini. Your team's rate limit is — Requests per Second (actual/limit): 0/8, Requests per Minute (actual/limit): 493/480. You can view your team rate limits at https://console.x.ai."}. Set `provide_traceback=True` for traceback.


Average Metric: 7555.00 / 10048 (75.2%):  85%|████████▌ | 10233/11970 [20:37<00:47, 36.58it/s]

2025/07/23 22:05:12 ERROR dspy.utils.parallelizer: Error for Example({'premise': "Gerald's banana that worries Pamela doesn't blacken.", 'hypothesis': 'Those women has exactly one banana that worries Pamela.', 'label': 'neutral'}) (input_keys={'hypothesis', 'premise'}): Both structured output format and JSON mode failed. Please choose a model that supports `response_format` argument. Original error: litellm.RateLimitError: RateLimitError: XaiException - Error code: 429 - {'code': 'Some resource has been exhausted', 'error': "Too many requests for team bc29b82a-19d1-4c03-acf7-cd8b768453be and model grok-3-mini. Your team's rate limit is — Requests per Second (actual/limit): 0/8, Requests per Minute (actual/limit): 493/480. You can view your team rate limits at https://console.x.ai."}. Set `provide_traceback=True` for traceback.


Average Metric: 7555.00 / 10048 (75.2%):  86%|████████▋ | 10327/11970 [20:37<00:21, 76.59it/s]

2025/07/23 22:05:12 ERROR dspy.utils.parallelizer: Error for Example({'premise': "Did those waitresses' shoe that hadn't annoyed the schools disappear?", 'hypothesis': "Roger have exactly one shoe that hadn't annoyed the schools.", 'label': 'neutral'}) (input_keys={'hypothesis', 'premise'}): Both structured output format and JSON mode failed. Please choose a model that supports `response_format` argument. Original error: litellm.RateLimitError: RateLimitError: XaiException - Error code: 429 - {'code': 'Some resource has been exhausted', 'error': "Too many requests for team bc29b82a-19d1-4c03-acf7-cd8b768453be and model grok-3-mini. Your team's rate limit is — Requests per Second (actual/limit): 0/8, Requests per Minute (actual/limit): 491/480. You can view your team rate limits at https://console.x.ai."}. Set `provide_traceback=True` for traceback.


Average Metric: 7555.00 / 10048 (75.2%):  89%|████████▉ | 10671/11970 [20:37<00:05, 225.01it/s]

2025/07/23 22:05:12 ERROR dspy.utils.parallelizer: Error for Example({'premise': "John's chair that aggravates Victoria didn't aggravate the school.", 'hypothesis': 'Kevin has exactly one chair that aggravates Victoria.', 'label': 'neutral'}) (input_keys={'hypothesis', 'premise'}): Both structured output format and JSON mode failed. Please choose a model that supports `response_format` argument. Original error: litellm.RateLimitError: RateLimitError: XaiException - Error code: 429 - {'code': 'Some resource has been exhausted', 'error': "Too many requests for team bc29b82a-19d1-4c03-acf7-cd8b768453be and model grok-3-mini. Your team's rate limit is — Requests per Second (actual/limit): 0/8, Requests per Minute (actual/limit): 489/480. You can view your team rate limits at https://console.x.ai."}. Set `provide_traceback=True` for traceback.


Average Metric: 7555.00 / 10048 (75.2%):  89%|████████▉ | 10685/11970 [20:37<00:05, 225.01it/s]

2025/07/23 22:05:12 ERROR dspy.utils.parallelizer: Error for Example({'premise': "If David's box that wasn't disgusting Marie did look like those prints, it's okay.", 'hypothesis': "Sara has exactly one box that wasn't disgusting Marie.", 'label': 'neutral'}) (input_keys={'hypothesis', 'premise'}): Both structured output format and JSON mode failed. Please choose a model that supports `response_format` argument. Original error: litellm.RateLimitError: RateLimitError: XaiException - Error code: 429 - {'code': 'Some resource has been exhausted', 'error': "Too many requests for team bc29b82a-19d1-4c03-acf7-cd8b768453be and model grok-3-mini. Your team's rate limit is — Requests per Second (actual/limit): 0/8, Requests per Minute (actual/limit): 490/480. You can view your team rate limits at https://console.x.ai."}. Set `provide_traceback=True` for traceback.


Average Metric: 7555.00 / 10048 (75.2%):  90%|█████████ | 10791/11970 [20:37<00:03, 310.36it/s]

2025/07/23 22:05:12 ERROR dspy.utils.parallelizer: Error for Example({'premise': "Donald's door that disgusts Sonia doesn't waste away.", 'hypothesis': 'Katherine has exactly one door that disgusts Sonia.', 'label': 'neutral'}) (input_keys={'hypothesis', 'premise'}): Both structured output format and JSON mode failed. Please choose a model that supports `response_format` argument. Original error: litellm.RateLimitError: RateLimitError: XaiException - Error code: 429 - {'code': 'Some resource has been exhausted', 'error': "Too many requests for team bc29b82a-19d1-4c03-acf7-cd8b768453be and model grok-3-mini. Your team's rate limit is — Requests per Second (actual/limit): 0/8, Requests per Minute (actual/limit): 490/480. You can view your team rate limits at https://console.x.ai."}. Set `provide_traceback=True` for traceback.


Average Metric: 7555.00 / 10048 (75.2%):  93%|█████████▎| 11131/11970 [20:37<00:01, 723.10it/s]

2025/07/23 22:05:13 ERROR dspy.utils.parallelizer: Error for Example({'premise': "Nina's commentary that embarrasses Sherry might have bored these senators.", 'hypothesis': 'Sandra has exactly one commentary that embarrasses Sherry.', 'label': 'neutral'}) (input_keys={'hypothesis', 'premise'}): Both structured output format and JSON mode failed. Please choose a model that supports `response_format` argument. Original error: litellm.RateLimitError: RateLimitError: XaiException - Error code: 429 - {'code': 'Some resource has been exhausted', 'error': "Too many requests for team bc29b82a-19d1-4c03-acf7-cd8b768453be and model grok-3-mini. Your team's rate limit is — Requests per Second (actual/limit): 0/8, Requests per Minute (actual/limit): 486/480. You can view your team rate limits at https://console.x.ai."}. Set `provide_traceback=True` for traceback.


Average Metric: 7555.00 / 10048 (75.2%):  93%|█████████▎| 11142/11970 [20:37<00:01, 723.10it/s]

2025/07/23 22:05:13 ERROR dspy.utils.parallelizer: Error for Example({'premise': "Gerald's banana that worries Pamela might blacken.", 'hypothesis': "Gerald's banana that worries Pamela does blacken.", 'label': 'neutral'}) (input_keys={'hypothesis', 'premise'}): Both structured output format and JSON mode failed. Please choose a model that supports `response_format` argument. Original error: litellm.RateLimitError: RateLimitError: XaiException - Error code: 429 - {'code': 'Some resource has been exhausted', 'error': "Too many requests for team bc29b82a-19d1-4c03-acf7-cd8b768453be and model grok-3-mini. Your team's rate limit is — Requests per Second (actual/limit): 0/8, Requests per Minute (actual/limit): 486/480. You can view your team rate limits at https://console.x.ai."}. Set `provide_traceback=True` for traceback.


Average Metric: 7555.00 / 10048 (75.2%):  93%|█████████▎| 11147/11970 [20:37<00:01, 723.10it/s]

2025/07/23 22:05:13 ERROR dspy.utils.parallelizer: Error for Example({'premise': "Sally's hat that has irritated Derek might have shrunk.", 'hypothesis': 'Winston Churchill has exactly one hat that has irritated Derek.', 'label': 'neutral'}) (input_keys={'hypothesis', 'premise'}): Both structured output format and JSON mode failed. Please choose a model that supports `response_format` argument. Original error: litellm.RateLimitError: RateLimitError: XaiException - Error code: 429 - {'code': 'Some resource has been exhausted', 'error': "Too many requests for team bc29b82a-19d1-4c03-acf7-cd8b768453be and model grok-3-mini. Your team's rate limit is — Requests per Second (actual/limit): 0/8, Requests per Minute (actual/limit): 487/480. You can view your team rate limits at https://console.x.ai."}. Set `provide_traceback=True` for traceback.


Average Metric: 7555.00 / 10048 (75.2%):  94%|█████████▍| 11228/11970 [20:38<00:01, 723.10it/s]

2025/07/23 22:05:13 ERROR dspy.utils.parallelizer: Error for Example({'premise': "Connie's commentary that irritated Deborah does worry James.", 'hypothesis': 'Todd has exactly one commentary that irritated Deborah.', 'label': 'neutral'}) (input_keys={'hypothesis', 'premise'}): Both structured output format and JSON mode failed. Please choose a model that supports `response_format` argument. Original error: litellm.RateLimitError: RateLimitError: XaiException - Error code: 429 - {'code': 'Some resource has been exhausted', 'error': "Too many requests for team bc29b82a-19d1-4c03-acf7-cd8b768453be and model grok-3-mini. Your team's rate limit is — Requests per Second (actual/limit): 0/8, Requests per Minute (actual/limit): 486/480. You can view your team rate limits at https://console.x.ai."}. Set `provide_traceback=True` for traceback.


Average Metric: 7555.00 / 10048 (75.2%):  94%|█████████▍| 11299/11970 [20:38<00:00, 816.24it/s]

2025/07/23 22:05:13 ERROR dspy.utils.parallelizer: Error for Example({'premise': "If Victoria's muffin that was impressing Mark did hurt Brian, it's okay.", 'hypothesis': 'Karla has exactly one muffin that was impressing Mark.', 'label': 'neutral'}) (input_keys={'hypothesis', 'premise'}): Both structured output format and JSON mode failed. Please choose a model that supports `response_format` argument. Original error: litellm.RateLimitError: RateLimitError: XaiException - Error code: 429 - {'code': 'Some resource has been exhausted', 'error': "Too many requests for team bc29b82a-19d1-4c03-acf7-cd8b768453be and model grok-3-mini. Your team's rate limit is — Requests per Second (actual/limit): 0/8, Requests per Minute (actual/limit): 486/480. You can view your team rate limits at https://console.x.ai."}. Set `provide_traceback=True` for traceback.


Average Metric: 7555.00 / 10048 (75.2%):  96%|█████████▋| 11549/11970 [20:38<00:45,  9.33it/s]  

2025/07/23 22:05:13 WARNING dspy.utils.parallelizer: Execution cancelled due to errors or interruption.


Exception: Execution cancelled due to errors or interruption.

In [ ]:
from dspy.teleprompt import MIPROv2

mipro = MIPROv2(metric=accuracy_metric, auto="light")
opted_mipro = mipro.compile(predictor, trainset=dev_ex)

In [ ]:
mipro_report = Evaluate(
        devset=test_ex,
        metric=accuracy_metric,
        return_outputs=True,
        num_threads=50,
        display_progress=True,
        display_table=False,
        provide_traceback=False
)(opted_mipro)

In [ ]:
mipro_score,mipro_report_results = mipro_report
print(f"Score: {mipro_score}")
mipro_test_pred = [label2id[out[1].label] for out in mipro_report_results]

In [ ]:
mipro_combined = compute_matrices(mipro_test_pred)

In [ ]:
display(pd.DataFrame(mipro_combined, index=["mipro_combined predictor"]))

In [ ]:
from dspy.teleprompt import Ensemble
ensemble = Ensemble(reduce_fn=dspy.majority)
combined = ensemble.compile([opted_rs, opted_mipro])

In [ ]:
combined_score, combined_report, combined_test_pred = evaluate(combined)
combined_combined = compute_matrices(combined_test_pred)
display(pd.DataFrame(combined_combined, index=["combined predictor"]))

In [42]:
from dspy.evaluate import Evaluate

test_results = {}
test_results_on_original_predicator = {}
for sec, examples in test_data.items():
    # print(f"Evaluating on test section: {sec}")
    # evaluator = Evaluate(
    #     devset=examples,
    #     metric=accuracy_metric,
    #     num_threads=20,
    #     display_progress=True,
    #     display_table=5,
    #     provide_traceback=True,
    #     max_errors=5,
    # )
    # result = evaluator(joint_predictor)
    # test_results[sec] = result
    print(f"Evaluating on test section:\t{sec}")
    evaluator = Evaluate(
        devset=examples,
        metric=accuracy_metric,
        return_outputs=True,
        num_threads=50,
        display_progress=True,
        display_table=False,
        provide_traceback=False
        # max_errors=30
    )
    eval_res = evaluator(joint_predictor)
    eval_res_on_original_predicator = evaluator(predictor)
    _, result_tuples = eval_res
    _, result_tuples_on_original_predicator = eval_res_on_original_predicator
    preds, refs = [], []
    for example, prediction, correct in result_tuples:
        if not hasattr(prediction, "label"):
            continue
        preds.append(prediction.label)
        refs.append(example.label)
    for example, prediction, correct in result_tuples_on_original_predicator:
        if not hasattr(prediction, "label"):
            continue
        preds.append(prediction.label)
        refs.append(example.label)
    test_results[sec] = {"preds": preds, "refs": refs}
    test_results_on_original_predicator[sec] = {"preds": preds, "refs": refs}

Evaluating on test section:	presupposition_all_n_presupposition
Average Metric: 565.00 / 570 (99.1%): 100%|██████████| 570/570 [00:00<00:00, 699.10it/s]

2025/07/20 19:59:51 INFO dspy.evaluate.evaluate: Average Metric: 565 / 570 (99.1%)



Average Metric: 540.00 / 570 (94.7%): 100%|██████████| 570/570 [00:00<00:00, 3489.60it/s]

2025/07/20 19:59:52 INFO dspy.evaluate.evaluate: Average Metric: 540 / 570 (94.7%)



Evaluating on test section:	presupposition_both_presupposition
Average Metric: 561.00 / 570 (98.4%): 100%|██████████| 570/570 [00:05<00:00, 96.67it/s] 

2025/07/20 20:00:02 INFO dspy.evaluate.evaluate: Average Metric: 561 / 570 (98.4%)



Average Metric: 557.00 / 570 (97.7%): 100%|██████████| 570/570 [00:00<00:00, 3310.46it/s]

2025/07/20 20:00:02 INFO dspy.evaluate.evaluate: Average Metric: 557 / 570 (97.7%)



Evaluating on test section:	presupposition_change_of_state
Average Metric: 318.00 / 570 (55.8%): 100%|██████████| 570/570 [00:00<00:00, 838.15it/s]

2025/07/20 20:00:11 INFO dspy.evaluate.evaluate: Average Metric: 318 / 570 (55.8%)



Average Metric: 321.00 / 570 (56.3%): 100%|██████████| 570/570 [00:00<00:00, 3965.14it/s]

2025/07/20 20:00:11 INFO dspy.evaluate.evaluate: Average Metric: 321 / 570 (56.3%)



Evaluating on test section:	presupposition_cleft_existence
Average Metric: 424.00 / 570 (74.4%): 100%|██████████| 570/570 [00:07<00:00, 72.57it/s] 

2025/07/20 20:00:21 INFO dspy.evaluate.evaluate: Average Metric: 424 / 570 (74.4%)



Average Metric: 412.00 / 570 (72.3%): 100%|██████████| 570/570 [00:00<00:00, 3830.72it/s]

2025/07/20 20:00:21 INFO dspy.evaluate.evaluate: Average Metric: 412 / 570 (72.3%)



Evaluating on test section:	presupposition_cleft_uniqueness
Average Metric: 283.00 / 570 (49.6%): 100%|██████████| 570/570 [00:03<00:00, 146.87it/s]

2025/07/20 20:00:30 INFO dspy.evaluate.evaluate: Average Metric: 283 / 570 (49.6%)



Average Metric: 283.00 / 570 (49.6%): 100%|██████████| 570/570 [00:00<00:00, 2733.48it/s]

2025/07/20 20:00:31 INFO dspy.evaluate.evaluate: Average Metric: 283 / 570 (49.6%)



Evaluating on test section:	presupposition_only_presupposition
Average Metric: 399.00 / 570 (70.0%): 100%|██████████| 570/570 [00:03<00:00, 181.41it/s]

2025/07/20 20:00:41 INFO dspy.evaluate.evaluate: Average Metric: 399 / 570 (70.0%)



Average Metric: 396.00 / 570 (69.5%): 100%|██████████| 570/570 [00:00<00:00, 2636.49it/s]


2025/07/20 20:00:42 INFO dspy.evaluate.evaluate: Average Metric: 396 / 570 (69.5%)


Evaluating on test section:	presupposition_possessed_definites_existence
Average Metric: 550.00 / 570 (96.5%): 100%|██████████| 570/570 [00:02<00:00, 202.21it/s]

2025/07/20 20:00:51 INFO dspy.evaluate.evaluate: Average Metric: 550 / 570 (96.5%)



Average Metric: 524.00 / 570 (91.9%): 100%|██████████| 570/570 [00:00<00:00, 3297.67it/s]

2025/07/20 20:00:52 INFO dspy.evaluate.evaluate: Average Metric: 524 / 570 (91.9%)



Evaluating on test section:	presupposition_possessed_definites_uniqueness
Average Metric: 264.00 / 570 (46.3%): 100%|██████████| 570/570 [00:07<00:00, 75.69it/s] 

2025/07/20 20:01:02 INFO dspy.evaluate.evaluate: Average Metric: 264 / 570 (46.3%)



Average Metric: 263.00 / 570 (46.1%): 100%|██████████| 570/570 [00:00<00:00, 2802.54it/s]


2025/07/20 20:01:03 INFO dspy.evaluate.evaluate: Average Metric: 263 / 570 (46.1%)


Evaluating on test section:	presupposition_question_presupposition
Average Metric: 492.00 / 570 (86.3%): 100%|██████████| 570/570 [00:10<00:00, 54.95it/s]

2025/07/20 20:01:14 INFO dspy.evaluate.evaluate: Average Metric: 492 / 570 (86.3%)



Average Metric: 496.00 / 570 (87.0%): 100%|██████████| 570/570 [00:00<00:00, 3431.60it/s]

2025/07/20 20:01:14 INFO dspy.evaluate.evaluate: Average Metric: 496 / 570 (87.0%)


In [38]:
metric_prf = combine(["precision", "recall", "f1"])
acc = load("accuracy")
label2id = {"entailment": 0, "neutral": 1, "contradiction": 2}

In [43]:
rows = []
all_preds, all_refs = [], []

for sec, res in test_results.items():
    print(f"Calculating metrics for section:\t{sec}")
    preds = [label2id[p] for p in res['preds']]
    refs = [label2id[ex] for ex in res['refs']]

    prf = metric_prf.compute(predictions=preds, references=refs, average="weighted")
    accuracy = acc.compute(predictions=preds, references=refs)["accuracy"]

    rows.append({"section": sec, "accuracy": accuracy, **prf})
    all_preds += preds
    all_refs += refs

overall_prf = metric_prf.compute(predictions=all_preds, references=all_refs, average="weighted")
overall_acc = acc.compute(predictions=all_preds, references=all_refs)["accuracy"]
rows.append({"section": "all", "accuracy": overall_acc, **overall_prf})

df_metrics = pd.DataFrame(rows)
display(df_metrics.set_index("section"))

Calculating metrics for section:	presupposition_all_n_presupposition
Calculating metrics for section:	presupposition_both_presupposition
Calculating metrics for section:	presupposition_change_of_state
Calculating metrics for section:	presupposition_cleft_existence
Calculating metrics for section:	presupposition_cleft_uniqueness
Calculating metrics for section:	presupposition_only_presupposition
Calculating metrics for section:	presupposition_possessed_definites_existence
Calculating metrics for section:	presupposition_possessed_definites_uniqueness
Calculating metrics for section:	presupposition_question_presupposition


accuracy  precision    recall  \
section                                                                        
presupposition_all_n_presupposition            0.969298   0.971071  0.969298   
presupposition_both_presupposition             0.980702   0.981264  0.980702   
presupposition_change_of_state                 0.560526   0.668705  0.560526   
presupposition_cleft_existence                 0.733333   0.832485  0.733333   
presupposition_cleft_uniqueness                0.496491   0.769661  0.496491   
presupposition_only_presupposition             0.697368   0.801856  0.697368   
presupposition_possessed_definites_existence   0.942105   0.945813  0.942105   
presupposition_possessed_definites_uniqueness  0.462281   0.763888  0.462281   
presupposition_question_presupposition         0.866667   0.878098  0.866667   
all                                            0.745419   0.826205  0.745419   

                                                     f1  
section                                                  
presupposition_all_n_presupposition            0.969294  
presupposition_both_presupposition             0.980716  
presupposition_change_of_state                 0.495187  
presupposition_cleft_existence                 0.725441  
presupposition_cleft_uniqueness                0.385187  
presupposition_only_presupposition             0.683522  
presupposition_possessed_definites_existence   0.942186  
presupposition_possessed_definites_uniqueness  0.342022  
presupposition_question_presupposition         0.864703  
all                                            0.738150

In [44]:
rows = []
all_preds, all_refs = [], []

for sec, res in test_results_on_original_predicator.items():
    print(f"Calculating metrics for section:\t{sec}")
    preds = [label2id[p] for p in res['preds']]
    refs = [label2id[ex] for ex in res['refs']]

    prf = metric_prf.compute(predictions=preds, references=refs, average="weighted")
    accuracy = acc.compute(predictions=preds, references=refs)["accuracy"]

    rows.append({"section": sec, "accuracy": accuracy, **prf})
    all_preds += preds
    all_refs += refs

overall_prf = metric_prf.compute(predictions=all_preds, references=all_refs, average="weighted")
overall_acc = acc.compute(predictions=all_preds, references=all_refs)["accuracy"]
rows.append({"section": "all", "accuracy": overall_acc, **overall_prf})

df_metrics = pd.DataFrame(rows)
display(df_metrics.set_index("section"))

Calculating metrics for section:	presupposition_all_n_presupposition
Calculating metrics for section:	presupposition_both_presupposition
Calculating metrics for section:	presupposition_change_of_state
Calculating metrics for section:	presupposition_cleft_existence
Calculating metrics for section:	presupposition_cleft_uniqueness
Calculating metrics for section:	presupposition_only_presupposition
Calculating metrics for section:	presupposition_possessed_definites_existence
Calculating metrics for section:	presupposition_possessed_definites_uniqueness
Calculating metrics for section:	presupposition_question_presupposition


accuracy  precision    recall  \
section                                                                        
presupposition_all_n_presupposition            0.969298   0.971071  0.969298   
presupposition_both_presupposition             0.980702   0.981264  0.980702   
presupposition_change_of_state                 0.560526   0.668705  0.560526   
presupposition_cleft_existence                 0.733333   0.832485  0.733333   
presupposition_cleft_uniqueness                0.496491   0.769661  0.496491   
presupposition_only_presupposition             0.697368   0.801856  0.697368   
presupposition_possessed_definites_existence   0.942105   0.945813  0.942105   
presupposition_possessed_definites_uniqueness  0.462281   0.763888  0.462281   
presupposition_question_presupposition         0.866667   0.878098  0.866667   
all                                            0.745419   0.826205  0.745419   

                                                     f1  
section                                                  
presupposition_all_n_presupposition            0.969294  
presupposition_both_presupposition             0.980716  
presupposition_change_of_state                 0.495187  
presupposition_cleft_existence                 0.725441  
presupposition_cleft_uniqueness                0.385187  
presupposition_only_presupposition             0.683522  
presupposition_possessed_definites_existence   0.942186  
presupposition_possessed_definites_uniqueness  0.342022  
presupposition_question_presupposition         0.864703  
all                                            0.738150

Let's examine the results:

| Section                                          | Accuracy | Precision | Recall   | F1 Score |
|--------------------------------------------------|----------|-----------|----------|----------|
| presupposition_all_n_presupposition              | 0.991228 | 0.991400  | 0.991228 | 0.991237 |
| presupposition_both_presupposition               | 0.984211 | 0.984519  | 0.984211 | 0.984191 |
| presupposition_change_of_state                   | 0.557895 | 0.652114  | 0.557895 | 0.491531 |
| presupposition_cleft_existence                   | 0.743860 | 0.835004  | 0.743860 | 0.736982 |
| presupposition_cleft_uniqueness                  | 0.496491 | 0.769661  | 0.496491 | 0.385733 |
| presupposition_only_presupposition               | 0.700000 | 0.813519  | 0.700000 | 0.685228 |
| presupposition_possessed_definites_existence     | 0.964912 | 0.965735  | 0.964912 | 0.964956 |
| presupposition_possessed_definites_uniqueness    | 0.463158 | 0.769068  | 0.463158 | 0.343951 |
| presupposition_question_presupposition           | 0.863158 | 0.875327  | 0.863158 | 0.861277 |
| all                                              | 0.751657 | 0.828935  | 0.751657 | 0.745117 |

Total F1 score of 0.745, not that much of an improvement :(

let's try to optimize it in another way.